source/common/http/filter/cors_filter.cc

```diff
@@ -0,0 +1,140 @@
+
+namespace Envoy {
+namespace Http {
+
+CorsFilter::CorsFilter(CorsFilterConfigConstSharedPtr config)
+    : config_(config), is_cors_request_(false) {}
+
+CorsFilter::~CorsFilter() {}
+
+FilterHeadersStatus CorsFilter::decodeHeaders(HeaderMap& headers, bool) {
+  initialize();
+
+  origin_ = headers.Origin();
+  if (origin_ == nullptr || origin_->value() == "") {
+    return FilterHeadersStatus::Continue;
+  }
+
+  is_cors_request_ = true;
+
+  const Http::HeaderEntry* method = headers.Method();
+  if (method != nullptr && method->value().c_str() != Http::Headers::get().MethodValues.Options) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (!cors_enabled_) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (allow_origin_ == "") {
+    return FilterHeadersStatus::Continue;
+  }
+
+  HeaderMapPtr response_headers{
+      new HeaderMapImpl{{Headers::get().Status, std::to_string(enumToInt(Http::Code::OK))}}};
+
+  if (allow_credentials_) {
+    response_headers->insertAccessControlAllowCredentials().value(
+        Http::Headers::get().CORSValues.True);
+    response_headers->insertAccessControlAllowOrigin().value(*origin_);
+  } else {
+    response_headers->insertAccessControlAllowOrigin().value(allow_origin_);
+  }
+
+  auto requestMethod = headers.AccessControlRequestMethod();
+  if (requestMethod == nullptr || requestMethod->value() == "") {
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (allow_methods_ != "") {
+    response_headers->insertAccessControlAllowMethods().value(allow_methods_);
+  }
+
+  if (allow_headers_ != "") {
+    response_headers->insertAccessControlAllowHeaders().value(allow_headers_);
+  }
+
+  if (expose_headers_ != "") {
+    response_headers->insertAccessControlExposeHeaders().value(expose_headers_);
+  }
+
+  if (max_age_ != "") {
+    response_headers->insertAccessControlMaxAge().value(max_age_);
+  }
+
+  response_headers->insertStatus().value(enumToInt(Http::Code::OK));
+  decoder_callbacks_->encodeHeaders(std::move(response_headers), true);
+
+  return FilterHeadersStatus::StopIteration;
+}
+
+FilterHeadersStatus CorsFilter::encodeHeaders(HeaderMap& headers, bool) {
+  if (!cors_enabled_ || !is_cors_request_) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (allow_credentials_) {
+    headers.insertAccessControlAllowCredentials().value(Http::Headers::get().CORSValues.True);
+    headers.insertAccessControlAllowOrigin().value(*origin_);
+  } else {
+    headers.insertAccessControlAllowOrigin().value(allow_origin_);
+  }
+
+  return FilterHeadersStatus::Continue;
+}
+
+void CorsFilter::setDecoderFilterCallbacks(StreamDecoderFilterCallbacks& callbacks) {
+  decoder_callbacks_ = &callbacks;
+};
+
+void CorsFilter::initialize() {
+  auto routeEntry = decoder_callbacks_->route()->routeEntry();
+  const Envoy::Router::CorsPolicy& routeCorsPolicy = routeEntry->corsPolicy();
+  const Envoy::Router::CorsPolicy& virtualHostCorsPolicy = routeEntry->virtualHost().corsPolicy();
```

Comment:
What about making these shared pointers to avoid all of the string copies?  You could instead have helpers allowOrigin() allowCredentials() etc. which would pull the appropriate value from the config with precedence.

Reply:
I was worrying about the copying. Good point. I'll make some changes. Having the helpers I could make the code cleaner and get rid of the initializer. The reason I didn't do it was that I thought calling it several times would maybe add too much overhead, but thinking about it, origin() would be called twice and others would be called only once so maybe not that much of a problem after all!